src/swarm/neo/client/mixins/SuspendableRequestCore.d

```diff
@@ -0,0 +1,1165 @@
+        MessageType = type of enum defining the types of messages used by this
+            request's protocol. Expected to contain the members Suspend, Resume,
+            Stop, and Ack (in addition to any others required by the request).
+
+*******************************************************************************/
+
+public struct SuspendableRequestControllerFiber ( Request, MessageType )
+{
+    import ocean.core.MessageFiber;
+
+    import swarm.neo.client.NotifierTypes;
+    import swarm.neo.client.RequestOnConn;
+    import swarm.neo.request.RequestEventDispatcher;
+
+    /// Request-on-conn event dispatcher.
+    private RequestOnConn.EventDispatcherAllNodes conn;
+
+    /// Request context.
+    private Request.Context* context;
+
+    /// Multi-fiber request event dispatcher.
+    private RequestEventDispatcher* request_event_dispatcher;
+
+    /// Fiber to suspend when waiting for I/O.
+    private MessageFiber fiber;
+
+    /***************************************************************************
+
+        Performs the logic of awaiting state change signals from the user-facing
+        controller class (see SuspendableController) and sending the
+        corresponding messages to the node.
+
+        Returns only when the request has been stopped by the user.
+
+        Params:
+            conn = request-on-conn event dispatcher
+            context = request context
+            request_event_dispatcher = multi-fiber request event dispatcher
+            fiber = fiber to suspend when waiting for I/O
+
+    ***************************************************************************/
+
+    public void handle ( RequestOnConn.EventDispatcherAllNodes conn,
+        Request.Context* context, RequestEventDispatcher* request_event_dispatcher,
+        MessageFiber fiber )
+    {
+        this.conn = conn;
+        this.context = context;
+        this.request_event_dispatcher = request_event_dispatcher;
+        this.fiber = fiber;
+
+        auto suspendable_control =
+            &this.context.shared_working.suspendable_control;
+
+        bool stopped, state_changed_in_notifier;
+        do
+        {
+            if ( !state_changed_in_notifier )
+            {
+                // Wait for user action.
+                auto event = this.request_event_dispatcher.nextEvent(this.fiber,
+                    Signal(suspendable_control.Signal.StateChangeRequested));
+                assert(event.active == event.active.signal);
+                assert(event.signal.code == suspendable_control.Signal.StateChangeRequested);
+            }
+
+            // Send state change message to node and wait for ACK.
+            stopped = this.changeRequestState();
+
+            // Notify the user that the state change has taken effect and handle
+            // the case where they request a new state change in the notifier.
+            state_changed_in_notifier = this.notifyUser(stopped);
+        }
+        while ( !stopped );
```

Comment:
Hmm wait... there's some conflict between this and https://github.com/gavin-norman-sociomantic/swarm-tsunami/blob/2ae0bb629d97d29219d121f5d569a93473a72355/test/neo/client/request/internal/GetAll.d#L413. I'll have to look into this. It looks like the same logic has been duplicated in two place, in slightly different forms.